### PR TITLE
Alert KubevirtVmHighMemoryUsage globally and not per VM

### PR DESCRIPTION
**Pull request by machadovilaca**
### What this PR does
Before this PR:

Some cluster had a lot of noise due to having many VMs with low free memory available

After this PR:

A single alert with the information of the number of VMs affected

jira-ticket: https://issues.redhat.com/browse/CNV-45063

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
Alert KubevirtVmHighMemoryUsage globally and not per VM
```

### CODE REVIEW

File: pkg/monitoring/rules/alerts/vms.go [hunk 1]
```diff
@@ -31,11 +31,11 @@ var (
...
 		{
 			Alert: "KubevirtVmHighMemoryUsage",
-			Expr:  intstr.FromString("kubevirt_vm_container_free_memory_bytes_based_on_working_set_bytes < 52428800 or kubevirt_vm_container_free_memory_bytes_based_on_rss < 52428800"),
+			Expr:  intstr.FromString("count(kubevirt_vm_container_free_memory_bytes_based_on_working_set_bytes < 52428800 or kubevirt_vm_container_free_memory_bytes_based_on_rss < 52428800)"),
 			For:   ptr.To(promv1.Duration("1m")),
 			Annotations: map[string]string{
-				"description": fmt.Sprintf("Container {{ $labels.container }} in pod {{ $labels.pod }} in namespace {{ $labels.namespace }} free memory is less than %s and it is close to requested memory", fiftyMB.String()),
-				"summary":     "VM is at risk of being evicted and in serious cases of memory exhaustion being terminated by the runtime.",
+				"description": fmt.Sprintf("There are {{ $value }} VMs where free memory is less than %s and it is close to requested memory", fiftyMB.String()),
+				"summary":     "VMs are at risk of being evicted and in serious cases of memory exhaustion being terminated by the runtime.",
```
Comment (iholder101) [inline pkg/monitoring/rules/alerts/vms.go:38]: Thanks @machadovilaca!

my 2cents:
* The term VM is a bit confusing here IMO. Perhaps we should emphasize that we're talking about the virt-launcher pod, not the guest.
* Is there a resolution that the admin / VM owner can see? i.e. using [AdditionalGuestMemoryOverheadRatio](https://github.com/kubevirt/kubevirt/blob/96c9f1180eb8770d58a1f74eb70f0a569d074e8e/staging/src/kubevirt.io/api/core/v1/types.go#L2495-L2501)?
* The JIRA ticket linked in the description is restricted to Red Hat employees only. I think it either has to be open for everyone or removed from the description.

Comment (machadovilaca) [reply pkg/monitoring/rules/alerts/vms.go:38]: > Thanks @machadovilaca!
> 
> my 2cents:
> 
> * The term VM is a bit confusing here IMO. Perhaps we should emphasize that we're talking about the virt-launcher pod, not the guest.

What do you think of keeping the VM term in the summary, but then being specific on the description, `X virt-laucher pods have less than 50Mi free memory ...`

> * Is there a resolution that the admin / VM owner can see? i.e. using [AdditionalGuestMemoryOverheadRatio](https://github.com/kubevirt/kubevirt/blob/96c9f1180eb8770d58a1f74eb70f0a569d074e8e/staging/src/kubevirt.io/api/core/v1/types.go#L2495-L2501)?

this is the runbook that is shared with the alert - https://github.com/kubevirt/monitoring/blob/main/docs/runbooks/KubevirtVmHighMemoryUsage.md

File: hack/prom-rule-ci/prom-rules-tests.yaml [hunk 2]
```diff
@@ -582,29 +582,35 @@ tests:
...
+        values: "19922944 18874368 18874368 17825792"
+
     alert_rule_test:
       - eval_time: 1m
         alertname: KubevirtVmHighMemoryUsage
         exp_alerts:
           - exp_annotations:
-              description: "Container compute in pod virt-launcher-testvm-123 in namespace ns-test free memory is less than 50Mi and it is close to requested memory"
-              summary: "VM is at risk of being evicted and in serious cases of memory exhaustion being terminated by the runtime."
+              description: "There are 2 VMs where free memory is less than 50Mi and it is close to requested memory"
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply hack/prom-rule-ci/prom-rules-tests.yaml:606]: I find the initial part: `There are 2 virt-launcher pods with 2 VMs` confusing

**Comment by sradco** [comment]
/hold
Please see my comment in the Jira issue.

**Pull request closed without merging**